src/main/scala/advent/solutions/Day2.scala

```diff
@@ -14,31 +16,103 @@ object Day2 {
       * @return The program after having been run on itself
       */
     def run(program: List[Int]): List[Int] = {
-      ???
+      val len = program.length - 1
+
+      @tailrec
+      def go(accum: List[Int], num: Int): List[Int] = {
+        if (num > len) accum else go(repair(num, accum), num + 4)
+      }
+
+      go(program, 0)
+    }
+
+    final case class Entry(index: Int, value: Int)
+
+    private val additionCode = 1
+    private val multiplicationCode = 2
+    private val lookupAndAdd = lookupAndOperate(_ + _)(_, _)
+    private val lookupAndMultiply = lookupAndOperate(_ * _)(_, _)
+
+    private def repair(initial: Int, program: List[Int]): List[Int] = {
+      if (program(initial) == additionCode)
+        repairProgram(program, lookupAndAdd(initial, program))
+      else if (program(initial) == multiplicationCode)
+        repairProgram(program, lookupAndMultiply(initial, program))
+      else program
+    }
+
+    private def lookupAndOperate(
+        f: (Int, Int) => Int
+    )(initial: Int, program: List[Int]): Entry = {
+      Entry(
+        program(initial + 3),
+        f(program(program(initial + 1)), program(program(initial + 2)))
+      )
+    }
+
+    private def repairProgram(
+        program: List[Int],
+        entry: Entry
+    ): List[Int] = {
+      program.updated(entry.index, entry.value)
     }
   }
 
   object Part2 {
 
+    import Day2.Part1.run
+
     /** Represents the two numbers provided at addresses 1 and 2 of an Intcode program */
     final case class Input(noun: Int, verb: Int)
 
     /** Calculates the input required to produce a given output
       *
       * @param program A list of opcodes and positions.  The positions at address 1 and 2 will be replaced with input
       * @param output  The given output of the program
-      *
       * @return The input that would be entered in the program to produce the given output
       */
-    def inputForOutput(program: List[Int], output: Int): Input = ???
+    def inputForOutput(program: List[Int], output: Int): Input = {
+
+      val nouns = LazyList.range(0, 100, 1)
+      val verbs = LazyList.range(0, 100, 1)
+      val inputs: LazyList[Input] =
+        nouns.flatMap(n => verbs.map(v => Input(n, v)))
+
+      def check(input: Input): Boolean = {
+        val modifiedProgram = generateUpdatedProgram(input, program)
+        val result = run(modifiedProgram)(0)
+        if (result == output) true else false
```

Comment:
`result == output` evaluates to `true`.  Instead of writing

```scala
if (result == output) true else false
```

you could write

```scala
result == output
```